### PR TITLE
Backing store

### PR DESCRIPTION
**Pull request by shans**

### CODE REVIEW

File: src/runtime/storageNG/direct-store.ts [hunk 1]
```diff
@@ -0,0 +1,241 @@
...
+  private version = 0;
+  private pendingException: Error | null = null;
+  private pendingResolves: Function[] = [];
+  private pendingRejects: Function[] = [];
+  private pendingDriverModels: {model: T['data']; version: number;}[] = [];
+  private state: DirectStoreState = DirectStoreState.Idle;
+  /*
+   * This class should only ever be constructed via the static construct method
+   */
+  private constructor(storageKey: StorageKey, exists: Exists, type: Type, mode: StorageMode, modelConstructor: new () => CRDTModel<T>) {
```
Comment (csilvestrini) [inline src/runtime/storageNG/direct-store.ts]: I think you can just delete this constructor, if it doesn't do anything but call its super constructor?

File: src/runtime/storageNG/direct-store.ts [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+      this.notifyIdle();
+    }
+  }
+  
+  private notifyIdle() {
+    if (this.pendingException) {
+      // this is termination.
+      this.pendingRejects.forEach(reject => reject(this.pendingException));
+    }
+    else {
```
Comment (csilvestrini) [inline src/runtime/storageNG/direct-store.ts]: Keep on one line

File: src/runtime/storageNG/direct-store.ts [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    if (state === DirectStoreState.Idle) {
+      // If we are already idle, this won't notify external parties.
+      this.notifyIdle();
+    }
+  }
+  
+  private notifyIdle() {
+    if (this.pendingException) {
+      // this is termination.
+      this.pendingRejects.forEach(reject => reject(this.pendingException));
```
Comment (csilvestrini) [inline src/runtime/storageNG/direct-store.ts]: Do you want to clear the list of pendingRejects after this? Or is that what the "this is termination" comment is referring to?

File: src/runtime/storageNG/direct-store.ts [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        //      this loop when a driver model arrives.
+        while (true) {
+          this.setState(DirectStoreState.AwaitingResponse);
+          // Work around a typescript compiler bug. Apparently typescript won't guarantee that
+          // a Map key you've just set will exist, but is happy to assure you that a private
+          // member variable couldn't possibly change in any function outside the local scope
+          // when within a switch statement. 
+          this.state = DirectStoreState.AwaitingResponse;
+          version += 1;
+          this.version = version;
```
Comment (csilvestrini) [inline src/runtime/storageNG/direct-store.ts]: How about: `this.version = ++version`

File: src/runtime/storageNG/direct-store.ts [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+            }
+            if (this.state !== DirectStoreState.AwaitingResponseDirty) {
+              // This shouldn't be possible as only a 'nack' should put us into
+              // AwaitingDriverModel, and only the above code should put us back
+              // into Idle.
+              throw new Error('reached impossible state in store state machine');
+            }
+            // fallthrough to re-execute the loop.
+          }
+          else {
```
Comment (csilvestrini) [inline src/runtime/storageNG/direct-store.ts]: Move else up to previous line

File: src/runtime/storageNG/direct-store.ts [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+              this.applyPendingDriverModels();
+              break;
+            }
+            if (this.state !== DirectStoreState.AwaitingResponseDirty) {
+              // This shouldn't be possible as only a 'nack' should put us into
+              // AwaitingDriverModel, and only the above code should put us back
+              // into Idle.
+              throw new Error('reached impossible state in store state machine');
+            }
+            // fallthrough to re-execute the loop.
```
Comment (csilvestrini) [inline src/runtime/storageNG/direct-store.ts]: Instead of comment, just say `continue;`

File: src/runtime/storageNG/direct-store.ts [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+      let noDriverSideChanges = true;
+      let theVersion = 0;
+      for (const {model, version} of models) {
+        try {
+          const {modelChange, otherChange} = this.localModel.merge(model);
+          this.deliverCallbacks(modelChange);
+          noDriverSideChanges = noDriverSideChanges && this.noDriverSideChanges(modelChange, otherChange, true);
+          theVersion = version;
+        }
+        catch (e) {
```
Comment (csilvestrini) [inline src/runtime/storageNG/direct-store.ts]: Same line

File: src/runtime/storageNG/direct-store.ts [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+  // Note that driver-side changes are stored in 'otherChange' when the merged operations/model is sent
+  // from the driver, and 'thisChange' when the merged operations/model is sent from a storageProxy.
+  // In the former case, we want to look at what has changed between what the driver sent us and what
+  // we now have. In the latter, the driver is only as up-to-date as our local model before we've
+  // applied the operations.
+  private noDriverSideChanges(thisChange: CRDTChange<T>, otherChange: CRDTChange<T>, messageFromDriver: boolean) {
+    if (messageFromDriver) {
+      return otherChange.changeType === ChangeType.Operations && otherChange.operations.length === 0;
+    }
+    else {
```
Comment (csilvestrini) [inline src/runtime/storageNG/direct-store.ts]: Same line

File: src/runtime/storageNG/backing-store.ts [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+  off(callback: number): void {
+    this.callbacks.delete(callback);
+  }
+  
+  async processStoreCallback(muxId: string, message: ProxyMessage<T>): Promise<boolean> {
+    return Promise.all([...this.callbacks.values()].map(callback => callback({mux_id: muxId, ...message}))).then(a => a.reduce((a, b) => a && b));
+  }
+
+  async onProxyMessage(message: ProxyMessage<T> & {mux_id: string}): Promise<boolean> {
+    if (this.stores[message.mux_id] == null) {
```
[the remark this answers is not in the document]
Comment (shans) [reply src/runtime/storageNG/backing-store.ts]: dictionaries don't have `has`

File: src/runtime/storageNG/backing-store.ts [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+      const store = await DirectStore.construct(this.storageKey.childWithComponent(message.mux_id), this.exists, this.type, this.mode, this.modelConstructor);
+      const id = store.on(msg => this.processStoreCallback(message.mux_id, msg));
+      this.stores[message.mux_id] = {store, id};
+    }
+    const {store, id} = this.stores[message.mux_id];
+    delete message.mux_id;
```
[the remark this answers is not in the document]
Comment (shans) [reply src/runtime/storageNG/backing-store.ts]: It wouldn't be a problem, but I like your suggestion of a separate argument (I had previously been attempting to shoehorn this into the standard Store interface but that didn't work)

File: src/runtime/storageNG/direct-store.ts (hunk 1 above)
Comment (shans) [reply src/runtime/storageNG/direct-store.ts]: This ensures that only DirectStore.construct can construct DirectStores, via the `private` keyword.

File: src/runtime/storageNG/direct-store.ts (hunk 3 above)
Comment (shans) [reply src/runtime/storageNG/direct-store.ts]: Yeah.

File: src/runtime/storageNG/direct-store.ts (hunk 6 above)
Comment (shans) [reply src/runtime/storageNG/direct-store.ts]: I think that would be less clear.

File: src/runtime/storageNG/direct-store.ts (hunk 7 above)
Comment (shans) [reply src/runtime/storageNG/direct-store.ts]: I have NFI how all these ended up on separate lines.

**Comment by shans** [comment]
PTAL @csilvestrini

File: src/runtime/storageNG/direct-store.ts (hunk 6 above)
Comment (csilvestrini) [reply src/runtime/storageNG/direct-store.ts]: `continue` says in code exactly what your comment says, except the compiler understands it too.

File: src/runtime/storageNG/backing-store.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (csilvestrini) [reply src/runtime/storageNG/backing-store.ts]: Oh right.. I got confused and thought it was a Map.

File: src/runtime/storageNG/direct-store.ts (hunk 1 above)
Comment (csilvestrini) [reply src/runtime/storageNG/direct-store.ts]: Gotcha

**Pull request merged**